File: 419 Battleships in a Board.java
/*
419. Battleships in a Board
https://leetcode.com/problems/battleships-in-a-board/

Given an 2D board, count how many battleships are in it. The battleships are represented with 'X's,
empty slots are represented with '.'s. You may assume the following rules:

You receive a valid board, made of only battleships or empty slots.

Battleships can only be placed horizontally or vertically. In other words,
they can only be made of the shape 1xN (1 row, N columns) or Nx1 (N rows, 1 column), where N can be of any size.
At least one horizontal or vertical cell separates between two battleships - there are no adjacent battleships.
Example:

X..X
...X
...X
In the above board there are 2 battleships.
Invalid Example:

...X
XXXX
...X
This is an invalid board that you will not receive - as battleships will always have a cell separating between them.
Follow up:
Could you do it in one-pass, using only O(1) extra memory and without modifying the value of the board?
*/
class Solution {
    public int countBattleships(char[][] board) {
        int count = 0;
        for(int i=0; i<board.length; ++i) {
            for(int j=0; j<board[0].length; ++j) {
                if(board[i][j]=='X' && (i == 0 || board[i-1][j]!='X') && (j ==0  || board[i][j-1]!='X')) {
                    ++count;
                }
            }
        }
        return count;
    }
}

File: 419 Battleships in a Board.py
'''
419. Battleships in a Board
https://leetcode.com/problems/battleships-in-a-board/

Given an m x n matrix board where each cell is a battleship 'X' or empty '.',
return the number of the battleships on board.

Battleships can only be placed horizontally or vertically on board.
In other words, they can only be made of the shape 1 x k (1 row, k columns) or
k x 1 (k rows, 1 column), where k can be of any size.
At least one horizontal or vertical cell separates between
two battleships (i.e., there are no adjacent battleships).

Example 1:

Input: board = [["X",".",".","X"],[".",".",".","X"],[".",".",".","X"]]
Output: 2

Example 2:

Input: board = [["."]]
Output: 0

Constraints:
    m == board.length
    n == board[i].length
    1 <= m, n <= 200
    board[i][j] is either '.' or 'X'.

Follow up: Could you do it in one-pass, using only O(1) extra memory and without modifying the values board?
'''
class Solution:
    def countBattleships(self, board: List[List[str]]) -> int:
        rowCount = len(board)
        colCount = len(board[0])
        count = 0
        for r in range(rowCount):
            for c in range(colCount):
                if board[r][c] == 'X' and (r == 0 or board[r-1][c] != 'X') and (c == 0 or board[r][c-1] != 'X'):
                    count += 1
        return count
